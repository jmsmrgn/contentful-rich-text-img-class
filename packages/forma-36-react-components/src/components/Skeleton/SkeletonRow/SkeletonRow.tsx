import React from 'react';

import TableRow from '../../Table/TableRow';
import SkeletonTableCell from './SkeletonTableCell/SkeletonTableCell';

export interface SkeletonRowProps {
  /** Defines the number of rows to be rendered */
  rowCount?: number;
  /** Defines the number of columns to be rendered */
  columnCount?: number;
}

const defaultProps = {
  rowCount: 1,
  columnCount: 5,
};

export const SkeletonRow = ({ rowCount, columnCount }: SkeletonRowProps) => {
  return (
    <>
      {Array.from(Array(rowCount)).map((_, rowIndex) => (
        <TableRow key={rowIndex}>
          {Array.from(Array(columnCount)).map((_, cellIndex) => (
            <SkeletonTableCell key={cellIndex} />
          ))}
        </TableRow>
      ))}
    </>
  );
};
SkeletonRow.defaultProps = defaultProps;

export default SkeletonRow;
