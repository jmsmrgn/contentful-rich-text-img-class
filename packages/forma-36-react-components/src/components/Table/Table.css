@import 'settings/colors';

.Table {
  width: 100%;
  border: 1px solid var(--color-element-light);
}
