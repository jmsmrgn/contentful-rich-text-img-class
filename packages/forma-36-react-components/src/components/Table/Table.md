### Usage

Use tables to present large amounts of data, and data which has multiple properties attached to it.

### Properties

#### Width

Tables always span the full width of the container they're in.

#### Elements

Tables always consist of a **table head**, describing the content of the respective columns, and the **table body**.

#### Rows & columns

Columns are ordered by relevance from left to right. Images, if present, have a higher priority, and should be placed in the first column. Checkboxes naturally have the highest priority.

#### Content sorting

Tables support sorting by columns (ascending and descending) by clicking on the column's table header. Even if the table doesn't support changing the sorting, it should show which column it's sorted by.

### Variations

All variations apply to the whole table, not single rows or cells.

#### Vertical padding

By default, cells have a vertical padding of 20px. If a table features images that are expected to exceed the height of the text content, the padding can be decreased to 10px to prevent the rows from taking up too much vertical space.

<img src="https://projects.invisionapp.com/static-signed/live-embed/118156364/191656271/1/latest/jjlEw5hiimk3lEQC5SElAlEQT5hL0luxKfBBPkf5nbhhMaVlEGEmsGNWNnkzZpozct8II2GsAmQSGHGu8XTDlE1SgfAlE/Padding-variations.png" alt="Vertical padding elements">

#### Single-line vs. multi-line

There are 2 options to deal with content that potentially exceeds the width of the table cells:

- **Single line cells:** Truncating the text automatically to fit into 1 line. The content is aligned to the vertical center of the cell.
- **Multi-line cells:** The text can take up multiple lines, although the length is capped at 200 characters. The content is aligned to the top of the cell. Use

<img src="https://projects.invisionapp.com/static-signed/live-embed/118156364/191656236/1/latest/5uV48Oi9vnNBlElNSlPVghDbTQlEY5bIDImeclETWv2mG9qKrA1YA16PXCqBu8emSaSCnnBiPl7pfo4p7NJSq8EglE/Line-variations.png" alt="Single-line vs. multi-line">

#### Table interactions

Generally, there are 2 ways tables rows can be interacted with:

- **Clickable rows:** the whole row is clickable. A column with a label can be added, in case it's not clear what action will be triggered when clicking a row.
- **Single action:** a single action, that's very specific and shouldn't be triggered by clicking the whole row, is placed in an extra column.
- **Multiple actions:** multiple actions can be placed together in an extra column.

Rows can have a checkbox for bulk actions (refer to the section below), which works independently from the interactions described above.

<img src="https://projects.invisionapp.com/static-signed/live-embed/118156364/191656209/1/latest/zJMkLlEPwlnYji05LOsMxamfJVNtIJ9iKJdbWAMnRo6l5ZQsgszZMlEPYjZM1lE1t1En3Wux4NqMpMUdqhAWxYduglE/Table-actions.png" alt="Table interactions" width="1100">

#### Sorting

If the table supports sorting, the respective columns are highlighted when being hovered over.

<img src="https://projects.invisionapp.com/static-signed/live-embed/118156364/192385196/1/latest/A9sNGLKfSlEVlE56knNM03v2621waKm5lEhnxHidf5XkzccT4XcssOlEzeVGWvReWSVcWRS3lPDxNj2eOuYW1NtPZQlE/Sorting.png" alt="Sorting">

#### Bulk actions

Upon selecting rows by clicking the checkbox, a bar displaying the actions available for the selected items appears below the table head. The selected rows are additionally highlighted with a light blue background. If a table spans multiple pages, accessible via pagination, the selected rows get unchecked automatically when switching to another page.

<img src="https://projects.invisionapp.com/static-signed/live-embed/118156364/191656191/1/latest/TSgwUaLvnq5idznSSy1iJkogGqwsJSmj8M8uJCD8SGlR8Vc6kt8qwIFtZZpnMfiZ05lEes2pUXuEmoIl02dvNnwlE/Bulk-actions.png" alt="Bulk actions">
