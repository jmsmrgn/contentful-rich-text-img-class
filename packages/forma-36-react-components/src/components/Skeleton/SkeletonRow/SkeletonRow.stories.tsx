import React from 'react';
import { storiesOf } from '@storybook/react';
import { number } from '@storybook/addon-knobs';

import SkeletonRow from './SkeletonRow';
import Table from '../../Table';
import TableHead from '../../Table/TableHead';
import TableBody from '../../Table/TableBody';
import TableCell from '../../Table/TableCell';
import TableRow from '../../Table/TableRow';

storiesOf('Components/Skeleton/SkeletonRow', module)
  .addParameters({
    propTypes: SkeletonRow['__docgenInfo'],
    component: SkeletonRow,
  })
  .add('default', () => (
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>Name</TableCell>
          <TableCell>Email</TableCell>
          <TableCell>Address</TableCell>
          <TableCell>Postal Code</TableCell>
          <TableCell>City</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <SkeletonRow
          rowCount={number('Number of rows', 5)}
          columnCount={number('Number of columns', 5)}
        />
      </TableBody>
    </Table>
  ));
