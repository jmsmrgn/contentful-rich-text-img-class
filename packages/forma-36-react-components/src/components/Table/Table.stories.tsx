import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, number, button } from '@storybook/addon-knobs';

import notes from './Table.md';
import Table from './Table';
import TableHead from './TableHead';
import TableBody from './TableBody';
import TableCell from './TableCell';
import TableRow from './TableRow';
import SkeletonRow from '../Skeleton/SkeletonRow';

function DefaultStory() {
  return (
    <div style={{ width: '800px' }}>
      <Table>
        <TableHead
          isSticky={boolean('isSticky', false)}
          offsetTop={text('offsetTop', '0px')}
        >
          <TableRow>
            <TableCell>Name</TableCell>
            <TableCell>Email</TableCell>
            <TableCell>Organization role</TableCell>
            <TableCell>Last activity</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell>Jane Roe</TableCell>
            <TableCell>jane@roe.com</TableCell>
            <TableCell>CEO</TableCell>
            <TableCell>August 29, 2018</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>John Doe</TableCell>
            <TableCell>john@doe.com</TableCell>
            <TableCell>CTO</TableCell>
            <TableCell>July 27, 2019</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </div>
  );
}

function WithLoadingStateStory() {
  const [isLoading, setIsLoading] = useState(true);

  button('toggle isLoading', () => setIsLoading(!isLoading));

  return (
    <div style={{ width: '800px' }}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Name</TableCell>
            <TableCell>Email</TableCell>
            <TableCell>Organization role</TableCell>
            <TableCell>Last activity</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {isLoading ? (
            <SkeletonRow
              rowCount={number('loadingRowCount', 4)}
              columnCount={number('loadingColumnCount', 4)}
            />
          ) : (
            <>
              <TableRow>
                <TableCell>Jane Roe</TableCell>
                <TableCell>jane@roe.com</TableCell>
                <TableCell>CEO</TableCell>
                <TableCell>August 29, 2018</TableCell>
              </TableRow>
              <TableRow>
                <TableCell>John Doe</TableCell>
                <TableCell>john@doe.com</TableCell>
                <TableCell>CTO</TableCell>
                <TableCell>July 27, 2019</TableCell>
              </TableRow>
            </>
          )}
        </TableBody>
      </Table>
    </div>
  );
}

storiesOf('Components/Table', module)
  .addParameters({
    propTypes: [
      Table['__docgenInfo'],
      TableHead['__docgenInfo'],
      TableBody['__docgenInfo'],
      TableCell['__docgenInfo'],
      TableRow['__docgenInfo'],
    ],
    component: Table,
    subcomponents: { TableHead, TableBody, TableCell, TableRow },
  })
  .add('default', () => <DefaultStory />, { notes })
  .add('with loading state', () => <WithLoadingStateStory />, { notes });
