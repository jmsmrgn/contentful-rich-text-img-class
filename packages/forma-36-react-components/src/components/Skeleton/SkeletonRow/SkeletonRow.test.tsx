import React from 'react';
import { shallow, mount } from 'enzyme';
import axe from '../../../utils/axeHelper';

import SkeletonRow from './SkeletonRow';
import Table from '../../Table';
import TableBody from '../../Table/TableBody';

it('renders the component', () => {
  const output = shallow(
    <Table>
      <TableBody>
        <SkeletonRow />
      </TableBody>
    </Table>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with 10 rows', () => {
  const output = shallow(
    <Table>
      <TableBody>
        <SkeletonRow rowCount={10} />
      </TableBody>
    </Table>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with 10 cells', () => {
  const output = shallow(
    <Table>
      <TableBody>
        <SkeletonRow columnCount={10} />
      </TableBody>
    </Table>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <Table>
      <TableBody>
        <SkeletonRow />
      </TableBody>
    </Table>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
