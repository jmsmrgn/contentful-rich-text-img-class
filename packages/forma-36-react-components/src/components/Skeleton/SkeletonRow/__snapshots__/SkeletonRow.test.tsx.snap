// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<table
  cellPadding="0"
  cellSpacing="0"
  className="Table"
  data-test-id="cf-ui-table"
>
  <TableBody
    testId="cf-ui-table-body"
  >
    <SkeletonRow
      columnCount={5}
      rowCount={1}
    />
  </TableBody>
</table>
`;

exports[`renders the component with 10 cells 1`] = `
<table
  cellPadding="0"
  cellSpacing="0"
  className="Table"
  data-test-id="cf-ui-table"
>
  <TableBody
    testId="cf-ui-table-body"
  >
    <SkeletonRow
      columnCount={10}
      rowCount={1}
    />
  </TableBody>
</table>
`;

exports[`renders the component with 10 rows 1`] = `
<table
  cellPadding="0"
  cellSpacing="0"
  className="Table"
  data-test-id="cf-ui-table"
>
  <TableBody
    testId="cf-ui-table-body"
  >
    <SkeletonRow
      columnCount={5}
      rowCount={10}
    />
  </TableBody>
</table>
`;
